primitives/core/src/bandersnatch.rs

```diff
@@ -0,0 +1,1016 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! VRFs backed by [Bandersnatch](https://neuromancer.sk/std/bls/Bandersnatch),
+//! an elliptic curve built over BLS12-381 scalar field.
+
+#[cfg(feature = "std")]
+use crate::crypto::Ss58Codec;
+use crate::crypto::{
+	ByteArray, CryptoType, CryptoTypeId, Derive, Public as TraitPublic, UncheckedFrom, VrfPublic,
+};
+#[cfg(feature = "full_crypto")]
+use crate::crypto::{DeriveError, DeriveJunction, Pair as TraitPair, SecretStringError, VrfSecret};
+
+use bandersnatch_vrfs::CanonicalSerialize;
+#[cfg(feature = "full_crypto")]
+use bandersnatch_vrfs::SecretKey;
+use codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+
+use sp_runtime_interface::pass_by::PassByInner;
+use sp_std::{boxed::Box, vec::Vec};
+
+/// Identifier used to match public keys against bandersnatch-vrf keys.
+pub const CRYPTO_ID: CryptoTypeId = CryptoTypeId(*b"band");
+
+/// Context used to produce a plain signature without any VRF input/output.
+#[cfg(feature = "full_crypto")]
+pub const SIGNING_CTX: &[u8] = b"SigningContext";
+
+// Max ring domain size.
+const RING_DOMAIN_SIZE: usize = 1024;
+
+#[cfg(feature = "full_crypto")]
+const SEED_SERIALIZED_LEN: usize = 32;
+
+// Short-Weierstrass form serialized sizes.
+const PUBLIC_SERIALIZED_LEN: usize = 33;
+const SIGNATURE_SERIALIZED_LEN: usize = 65;
+const PREOUT_SERIALIZED_LEN: usize = 33;
+const PEDERSEN_SIGNATURE_SERIALIZED_LEN: usize = 163;
+const RING_PROOF_SERIALIZED_LEN: usize = 592;
+
+// Max size of serialized ring-vrf context params.
+//
+// This size is dependent on the ring size.
+//
+// Some values:
+//  ring_size → ~serialized_size
+//   512        →  74 KB
+//  1024        → 147 KB
+//  2048        → 295 KB
+// NOTE: This is quite big but looks like there is an upcoming fix
+// in the backend.
+const RING_CONTEXT_SERIALIZED_LEN: usize = 147752;
```

Comment:
@skunert Good question. I was wondering this myself. (: 

If this is supposed to be 147KB because the max supported ring size is 1024 then it's either too small (147 * 1024 = 150528) or too big (147 * 1000 = 147000) or the serialized sizes in the comment are just approximations. Maybe @davxy you could add an extra comment how this was calculated and why?

Reply:
This is the empirical result and also the result of `RingContext::max_encoded_len`.
The w3f guys told be that probably this size can be reduced. But I'll keep this as an internal detail and at the current stage it doesn't matter a lot